File: 0146-lru-cache/0146-lru-cache.py
## time : o(1)
## space : o(n)
class LRUCache:
    class Node:
        def __init__(self, key, val):
            self.key = key
            self.val = val
            self.next = None
            self.prev = None

    def __init__(self, capacity):
        self.cap = capacity
        self.head = self.Node(-1, -1)
        self.tail = self.Node(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head
        self.cache = {}

    def add_node(self, new_node):
        temp = self.head.next
        new_node.next = temp
        new_node.prev = self.head
        self.head.next = new_node
        temp.prev = new_node

    def delete_node(self, del_node):
        del_prev = del_node.prev
        del_next = del_node.next
        del_prev.next = del_next
        del_next.prev = del_prev

    def get(self, key_):
        if key_ in self.cache:
            res_node = self.cache[key_]
            res = res_node.val
            del self.cache[key_]
            self.delete_node(res_node)
            self.add_node(res_node)
            self.cache[key_] = self.head.next
            return res
        return -1

    def put(self, key_, value):
        if key_ in self.cache:
            existing_node = self.cache[key_]
            del self.cache[key_]
            self.delete_node(existing_node)

        if len(self.cache) == self.cap:
            del_key = self.tail.prev.key
            del_node = self.tail.prev
            del self.cache[del_key]
            self.delete_node(del_node)

        new_node = self.Node(key_, value)
        self.add_node(new_node)
        self.cache[key_] = self.head.next


# Example usage:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key, value)

File: 0146-lru-cache/0146-lru-cache.cpp
class LRUCache {
public:
    // doubly linked list class
    class node {
      public:
        int key;
        int val;
        node * next;
        node * prev;
      // contructor  
      node(int _key, int _val) {
        key = _key;
        val = _val;
      }
    };
    
    // creating a head, tail and assign dummy values (-1, -1)
    node* head = new node(-1, -1);
    node* tail = new node(-1, -1);
    
    // capacity of cache
    int cap;
    // hasmap
    unordered_map<int, node*>m;
    
    
    LRUCache(int capacity) {
        // assign
        cap = capacity;
        head->next = tail;
        tail->prev = head;
    }
    
    //add node
    void addnode(node* newnode){
        node* temp = head->next;
        newnode->next = temp;
        newnode->prev = head;
        head->next = newnode;
        temp->prev = newnode;
        
    }
    
    // delete node
    void deletenode(node* delnode){
        node* delprev = delnode->prev;
        node* delnext = delnode->next;
        delprev->next = delnext;
        delnext->prev = delprev;
    }
    
    int get(int key_) {
        // first check if the key existed in cache i,e hashmap
        if(m.find(key_) != m.end()){
            // if the key exists in the map then take node address and store in hashmap by m[key_]
            node* resnode = m[key_];
            // the value will be of the node
            int res = resnode->val;
            // once done the above step delete the address because its the old adress
            m.erase(key_);
            deletenode(resnode);
            addnode(resnode);
            // add will be right after head so m[key_] will be new address
            m[key_] = head->next;
            return res;
        }
        return -1;
    }
    
    void put(int key_, int value) {
        // first check if the key existed in cache i,e hashmap
        if(m.find(key_) != m.end()){
            // if exists get the existing address
            node* existingnode = m[key_];
            m.erase(key_);
            deletenode(existingnode);
        }
        // if th size is full delete the least recently used
        if(m.size() == cap){
            // least recently used is tails previous
            m.erase(tail->prev->key);
            deletenode(tail->prev);
        }
        // add the node
        addnode(new node(key_, value));
        m[key_] = head->next;
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */